experimental/framework/fluid-llm/src/test/object-diff/objectDiffArray.spec.ts

```diff
@@ -0,0 +1,247 @@
+import { strict as assert } from "node:assert";
+
+import { objectDiff, type Difference } from "../../object-diff/index.js";
+
+describe("objectDiff - arrays", () => {
+	it("top level array & array diff", () => {
+		assert.deepStrictEqual(objectDiff(["test", "testing"], ["test"]), [
+			{
+				type: "REMOVE",
+				path: [1],
+				oldValue: "testing",
+			},
+		]);
+	});
+
+	it("nested array", () => {
+		assert.deepStrictEqual(objectDiff(["test", ["test"]], ["test", ["test", "test2"]]), [
+			{
+				type: "CREATE",
+				path: [1, 1],
+				value: "test2",
+			},
+		]);
+	});
+
+	it("object in array in object", () => {
+		assert.deepStrictEqual(
+			objectDiff({ test: ["test", { test: true }] }, { test: ["test", { test: false }] }),
+			[
+				{
+					type: "CHANGE",
+					path: ["test", 1, "test"],
+					value: false,
+					oldValue: true,
+				},
+			],
+		);
+	});
+
+	it("array to object", () => {
+		assert.deepStrictEqual(objectDiff({ data: [] }, { data: { val: "test" } }), [
+			{ type: "CHANGE", path: ["data"], value: { val: "test" }, oldValue: [] },
+		]);
+	});
+});
+
+describe("objectDiff - arrays with object ID's", () => {
+	it("object with id is moved from a new deleted array index", () => {
+		const diffs: Difference[] = objectDiff(
+			{ state: ["test", { id: "1", test: true }] },
+			{ state: [{ id: "1", test: true }] },
+			{
+				cyclesFix: true,
+				useObjectIds: {
+					idAttributeName: "id",
+				},
+			},
+		);
+
+		assert.deepStrictEqual(diffs, [
+			{
+				type: "CHANGE",
+				path: ["state", 0],
+				value: { id: "1", test: true },
+				oldValue: "test",
+			},
+			{
+				type: "MOVE",
+				path: ["state", 1],
+				newIndex: 0,
+				value: { id: "1", test: true },
+			},
+		]);
+	});
+
+	it("objects with id swap array indexes", () => {
+		const diffs: Difference[] = objectDiff(
+			{
+				state: [
+					{ id: "1", test: true },
+					{ id: "2", test: true },
+				],
+			},
+			{
+				state: [
+					{ id: "2", test: true },
+					{ id: "1", test: true },
+				],
+			},
+			{
+				cyclesFix: true,
+				useObjectIds: {
+					idAttributeName: "id",
+				},
+			},
+		);
+
+		assert.deepStrictEqual(diffs, [
+			{
+				type: "MOVE",
+				path: ["state", 0],
+				value: { id: "1", test: true },
+				newIndex: 1,
+			},
+			{
+				type: "MOVE",
+				path: ["state", 1],
+				value: { id: "2", test: true },
+				newIndex: 0,
+			},
+		]);
```

Comment:
As we discussed in the AI sync meeting, I'm a bit iffy about this. If the code understands that having these two operations is "really a single operation" (i.e. once you move the first item to the second position, there's no need for more moves), it might be ok, just a bit confusing. But I suspect applying these set of diffs does not actually produce the end state that the test says it should.

Reply:
Curious why this isn't `REMOVE` operation.

Original: `{ state: ["test", { id: "1", test: true }] }`
Result: `{ state: [{ id: "1", test: true }] }`

```json
{
    type: "REMOVE",
    path: ["state", 0], // treeNode[state][0]
    value: "test",
}
```

Is there a reason that the `sharedTreeDiff()` is dissecting this as two-moves?